src/each/each.py

```diff
@@ -24,17 +27,164 @@
+    def exists(self):
+        """Whether or not this work item still exists."""
+
+    @abstractmethod
+    def as_input_file(self):
+        """This work item as a file descriptor.
+
+        This file descriptor is used as STDIN on a user-provided command.
+        """
+
+    @abstractmethod
+    def as_argument(self):
+        """This work item as a command-line argument."""
+
+    @abstractmethod
+    def write_in_file(self, path):
+        """Create a file at 'path' that contains the input data for this work item."""
+
+
+@attr.s()
+class FileWorkItem(WorkItem):
+    """A file to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The location of the file on disk."""
+    path = attr.ib()
+
+    def exists(self):
+        """A file work item exists only if the file exists."""
+        return os.path.exists(self.path)
+
+    def as_input_file(self):
+        return os.open(self.path, os.O_RDONLY)
+
+    def as_argument(self):
+        return os.path.abspath(self.path)
+
+    def write_in_file(self, path):
+        """The ``in`` file is a symlink to the original file."""
+        os.symlink(os.path.abspath(self.path), path)
+
+
+@attr.s()
+class LineWorkItem(WorkItem):
+    """A line to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The line itself, including line ending, as text."""
+    line = attr.ib()
+
+    def exists(self):
+        """Whether or not this work item still exists.
+
+        A line always exists.
+        """
+        return True
+
+    def as_input_file(self):
+        r, w = os.pipe()
+        os.write(w, self.line.encode("utf-8"))
+        return r
+
+    def as_argument(self):
+        return self.line.rstrip("\n")
+
+    def write_in_file(self, path):
+        """The ``in`` file for a line work item is a file with just that line."""
+        with open(path, "w") as in_file:
+            in_file.write(self.line)
+
+
+def work_items_from_path(path):
+    """Load work items from a user-supplied path.
+
+    If 'path' is a directory, our work items are files. If it's a # file, our
+    work items are lines.
+    """
+    try:
+        return work_items_from_directory(path)
+    except NotADirectoryError:
+        with open(path, "r") as stream:
+            return work_items_from_file(stream)
+
+
+def work_items_from_directory(path):
+    """Get a list of work items from a directory.
+
+    Each file in the directory is a work item.
+    """
+    return (FileWorkItem(name=s, path=os.path.join(path, s)) for s in os.listdir(path))
+
+
+"""What we consider a 'simple name'.
+
+These are appended to the hashes of lines when generating output,
+as a convenience to the humans who must contend with it.
+"""
+simple_name_re = re.compile("^[A-Za-z0-9_-]+$")
+
+
+"""The longest simple name we allow."""
+MAX_NAME_LENGTH = 100
+
+
+def work_items_from_file(stream):
```

Comment:
nit: Would `work_items_from_lines` be a better name? Makes it clear what the file contents to work item mapping is.

Reply:
`work_items_from_file` name is from a previous factoring